Fixed Issue #7460: Raised an error if argument to xlim is invalid

nvedant07 opened the pull request:
Fixed issue #7460

NelleV commented:
Can you please add a test ?

WeatherGod commented on lib/matplotlib/axes/_base.py:
```diff
@@ -2871,8 +2871,12 @@ def set_xlim(self, left=None, right=None, emit=True, auto=False, **kw):
 
         self._process_unit_info(xdata=(left, right))
         if left is not None:
+            if isinstance(left,float) and (math.isnan(left) or np.isinf(left)):
```
is this going to work if passed a numpy float?

phobson replied:
perhaps `numpy.isreal` and `numpy.isfinite` would be better to use here.

NelleV replied:
I think it still works, but I think we should leverage as much as possible numpy considering it is one of our dependencies.

WeatherGod replied:
think of it this way. We recently discovered some bugs that were because numpy ints and python ints aren't the same in py3k, and so `isinstance(a, int)` of a numpy int wouldn't work anymore. Better safe than sorry.

WeatherGod commented:
what about `set_ylim()`? Also, these methods are duplicated in `lib/mpl_toolkits/mplot3d/axes3d.py` due to subtle differences between the 2d and 3d versions. This check should also go in there as well.

tacaswell commented:
Something has gone _very_ wring with git here....What were you trying to do?

Probably worth jumping in https://gitter.im/matplotlib/matplotlib to get help sorting this out.

nvedant07 commented:
I updated my local fork's master to be in sync with the upstream : matplotlib/matplotlib (master) and then force pushed the master to my local fork's branch(which is linked to this PR). I guess that is what made things go very wrong. Pruning isn't working. What do you suggest i do?

WeatherGod commented:
Don't push master to the branch -- you do rebases instead, or at least do
merges (but we prefer rebases). Never do pushes like that.

If you jump onto gitter, some very savy git users can use various reflog
tricks to hopefully fix your branch.

On Wed, Mar 1, 2017 at 1:41 AM, Vedant Nanda <notifications@github.com>
wrote:

> I updated my local fork's master to be in sync with the upstream :
> matplotlib/matplotlib (master) and then force pushed the master to my local
> fork's branch(which is linked to this PR). I guess that is what made things
> go very wrong. Pruning isn't working. What do you suggest i do?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/8022#issuecomment-283258551>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AARy-JDaHXoGBzc-7yZxhcFIy8p24IhHks5rhRMegaJpZM4L3jpJ>
> .
>

nvedant07 commented:
i have reverted the merge which led to all this. Now the branch is in the right state, however it shows 226 commits. Perhaps best to close this PR and open a new one?

WeatherGod commented:
no... do an interactive rebase to get rid of them all. It is called rebase
and squash.

On Wed, Mar 1, 2017 at 1:36 PM, Vedant Nanda <notifications@github.com>
wrote:

> i have reverted the merge which led to all this. Now the branch is in the
> right state, however it shows 226 commits. Perhaps best to close this PR
> and open a new one?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/8022#issuecomment-283428510>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AARy-AwUusy7aj1Z1MGaJ-Xsfh3LGfRWks5rhbqfgaJpZM4L3jpJ>
> .
>

nvedant07 commented:
Okay done, thanks a lot for the help.
I have added the tests and also put the checks in `set_ylim()` and in all of `set_xlim3d()` `set_ylim3d()` and `set_zlim3d()` in `lib/mpl_toolkits/mplot3d/axes3d.py`.

WeatherGod commented on lib/mpl_toolkits/tests/test_mplot3d.py:
```diff
@@ -491,3 +491,34 @@ def test_autoscale():
+    with pytest.raises(ValueError) as err:
+        plt.set_xlim3d(right=np.nan)
+    with pytest.raises(ValueError) as err:
+        plt.set_xlim3d(right=np.inf)
```
These won't work because the default axes object do not have these methods. You will need to explicitly create the 3d axes object and call the methods on it.

WeatherGod commented on lib/mpl_toolkits/tests/test_mplot3d.py:
```diff
@@ -491,3 +491,34 @@ def test_autoscale():
+    with pytest.raises(ValueError) as err:
+        plt.set_zlim3d(top=np.inf)
+
+    err.match('NaN or Inf cannot be the argument values')
```
I am not an expert in pytest, but I am not quite sure this is correct. Wouldn't the `err` object get clobbered in each test, so you are really only comparing with the last one?

WeatherGod commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -4974,23 +4974,30 @@ def test_bar_single_height():
     with pytest.raises(ValueError) as err:
-        plt.set_xlim(left=np.nan)
+        plt.xlim(left=np.nan)
+    assert err.value.message == error_string
```
Do exceptions have this `message` attribute in py3k? The test suite is failing for py3k, saying that it doesn't (but it works for python27).

phobson replied:
along those lines, I'm comfortable with simpley checking the error type

nvedant07 commented:
@phobson @WeatherGod  I've added tests that simply check error type. Travis and Appveyor both build successfully now. I hope this is fine.

phobson commented on lib/matplotlib/axes/_base.py:
```diff
@@ -2874,8 +2874,14 @@ def set_xlim(self, left=None, right=None, emit=True, auto=False, **kw):
 
         self._process_unit_info(xdata=(left, right))
         if left is not None:
+            if (isinstance(left, float) and
```
this is code block is repeater several times throughout the PR.

i recommend factoring it out into it's own method. something like:

```python
...
class _AxesBase(martist.Artist):
    ...
    def _validate_axis_limit(limit):
        if limit is not None:
            if isinstance(limit, float) and
                    (not np.isreal(limit) or not np.isfinite(right)):
                raise ValueError(...)
            return self.convert_xunits(limit)
```

and then all of these blocks can just be, e.g.
```python
right = self._validate_axis_limit(right)
```

phobson commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -4972,3 +4972,23 @@ def test_bar_single_height():
     ax.bar(0, 1, bottom=range(4), width=1, orientation='horizontal')
+
+
+def test_invalid_axes_limits():
```
I think you should try to parametrize the tests too:

a first pass would be something as simple as:

```python
@pytest.mark.parametrize('value', [np.inf, np.nan])
def test_invalid_axes_limits(value)
    with pytest.raises(ValueError):
        plt.xlim(left=value)

    with pytest.raises(ValueError):
        plt.xlim(right=value)

    with pytest.raises(ValueError):
        plt.ylim(bottom=value)

    with pytest.raises(ValueError):
        plt.ylim(top=value)
```

Going two steps further, you could parametrize the limit functions and sides, but that might be going a bit too far:

```python
@pytest.mark.parametrize('value', [np.inf, np.nan])
@pytest.mark.parametrize(('key', 'setter'), [
    (plt.xlim, 'left'),
    (plt.xlim, 'right'),
    (plt.ylim, 'bottom'),
    (plt.ylim, 'top'),
])
def test_invalid_axes_limits(key, setter, value)
    limit = {key: value}
    with pytest.raises(ValueError):
        setter(**limit)
```

QuLogic replied:
I think you might have the 'side' and 'setter' swapped in the `parametrize` call.

phobson replied:
@QuLogic yup I'll fix

QuLogic commented:
Going back a few commits, using `err.match` was fine; what @WeatherGod was saying is that you needed to repeat it for _every_ exception context, not just once at the end. But if you parametrize the test like @phobson asks, then you can put back the `err.match` and only need to do it once.

Also, please rebase instead of merging.

tacaswell commented:
#7744 fixed the bug.  Can this PR be re-based and sort out if the tests should be expanded?

nvedant07 commented:
As @phobson suggested, i have parametrized the tests and also moved the repeated code inside a function.
@tacaswell I looked at #7744 , it puts the check only in `lib/matplotlib/axes/_base.py` (as a result i have removed my checks from there). I have also put these checks in `lib/mpl_toolkits/mplot3d/axes3d.py` and the corresponding tests in `lib/mpl_toolkit/tests/test_mplot3d.py`  as @WeatherGod suggested a few commits back.

QuLogic commented on lib/mpl_toolkits/tests/test_mplot3d.py:
```diff
@@ -491,3 +491,18 @@ def test_autoscale():
+
+@pytest.mark.parametrize('value', [np.inf, np.nan])
+@pytest.mark.parametrize(('setter', 'side'), [
+    (plt.figure().add_subplot(111, projection='3d').set_xlim3d, 'left'),
```
Cleanup will work on the test, but I don't think it will work right on the parametrization. Consequently, I don't think it makes sense to create the figure in this list. It should be created in the test itself and only the setter name should be here. You can use `getattr` to dynamically access the method you want.

phobson commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -604,8 +615,10 @@ def set_xlim3d(self, left=None, right=None, emit=True, auto=False, **kw):
         self._process_unit_info(xdata=(left, right))
         if left is not None:
+            self._validate_axis_limits(left)
             left = self.convert_xunits(left)
```
I still recommend moving the `convert_xunits` into your `_validate...` method

nvedant07 commented:
@phobson @QuLogic I've made the changes and squashed the commits into one. Please take a look

tacaswell commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -585,6 +585,19 @@ def _determine_lims(self, xmin=None, xmax=None, *args, **kwargs):
+            if (isinstance(limit, float) and
+                    (not np.isreal(limit) or not np.isfinite(limit))):
+                raise ValueError("NaN or Inf cannot be the argument values")
+            return convert(limit)
```
might want to do the conversion first in case that causes it to be invalid?

QuLogic commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -585,6 +585,19 @@ def _determine_lims(self, xmin=None, xmax=None, *args, **kwargs):
 
+    def _validate_axis_limits(self, limit, convert):
+        """
+        If the axis limits being set are infinite, this function
```
First line should be a full sentence; can be shortened to "Raise if specified axis limits are infinite."

QuLogic commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -585,6 +585,19 @@ def _determine_lims(self, xmin=None, xmax=None, *args, **kwargs):
+        if limit is not None:
+            if (isinstance(limit, float) and
+                    (not np.isreal(limit) or not np.isfinite(limit))):
+                raise ValueError("NaN or Inf cannot be the argument values")
```
Users aren't going to know what "argument values" means; should say "Axis limits cannot be ..."

QuLogic commented on lib/mpl_toolkits/tests/test_mplot3d.py:
```diff
@@ -513,3 +513,20 @@ def test_autoscale():
+    with pytest.raises(ValueError):
+        fig = plt.figure()
+        obj = fig.add_subplot(111, projection='3d')
+        getattr(obj, setter)(**limit)
```
Only this line should be in the context, just in case the first two raise other things accidentally.

nvedant07 commented:
@QuLogic @tacaswell made the suggested changes and rebased

QuLogic commented:
One _really_ minor thing, but I think the message on the first commit might have been messed up. It's talking about the tests and a cleaner version of your work, which seems like a message for a second (now squashed) commit, and maybe the first commit message was lost?

nvedant07 commented:
@QuLogic updated the commit messages

QuLogic commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -587,6 +587,18 @@ def _determine_lims(self, xmin=None, xmax=None, *args, **kwargs):
+        if limit is not None:
+            converted_limits = convert(limit)
+            if (isinstance(limit, float) and
+                    (not np.isreal(limit) or not np.isfinite(limit))):
```
Oh bother, I just noticed this should probably be checking the converted limits. Do you want to open a new PR for that, @nvedant07?

nvedant07 replied:
Sure , i have opened a new PR for this - #8474
